crates/ra_ide/src/references.rs

```diff
@@ -210,11 +224,50 @@ fn process_definition(
     refs
 }
 
+fn access_mode(kind: NameKind, name_ref: &ast::NameRef) -> Option<ReferenceAccess> {
+    match kind {
+        NameKind::Local(_) | NameKind::Field(_) => {
+            //LetExpr or BinExpr
+            name_ref.syntax().ancestors().find_map(|node| {
+                match_ast! {
+                    match (node) {
+                        ast::BinExpr(expr) => {
```

Comment:
If we only look at `BinExpr`, it's simpler to `.filter_map(ast::BinExpr::cast)`

Reply:
`LetExpr` is on the TODO list and I should update the comment above. Does `filter_map` still make  sense? I think I'd just have to change it later.

Reply:
Hm, yeah, `match_ast` makes sense here. That said, the rightward drift seems non-trivial here. Perhaps  something can be factored into a function? 

Or maybe we can just use an early return for the outer match?

Reply:
In thinking about it... if we're in a `ast::BinExpr` and not in the lhs of an assignment we can guarantee we're in a read so I think I can change that inner bit to:

```rust
 ast::BinExpr(expr) => {
	if expr.op_kind()?.is_assignment() {
		// If the variable or field ends on the LHS's end then it's a Write (covers fields and locals).
		// FIXME: This is not terribly accurate.
		if let Some(lhs) = expr.lhs() {
			if lhs.syntax().text_range().end() == name_ref.syntax().text_range().end() {
				return Some(ReferenceAccess::Write);
			}
		}
	}
	
	return Some(ReferenceAccess::Read);
},
```